server/ServerApp.cpp

```diff
@@ -3371,6 +3372,14 @@ void ServerApp::PostCombatProcessTurns() {
     m_universe.BackPropagateObjectMeters();
     empires.BackPropagateMeters();
 
+    // add new techs
+    for (const auto& empire_techs : empires_innovations) {
```

Comment:
What's the reason for adding techs to empires here / now, though?

It's moved from before scripted turn events, a meter update, pop growth production research phase (which currently just does planet depopulation), and clamping and back-propagation, to after all of those. What specifically is it being moved after to do something usefully different? I'm guessing it's the `ApplyMeterEffectsAndUpdateMeters` call? But why specifically?

Adding the techs is also still before an empire capital check, various visibility and object knowledge tests and empire elimination... Why not later? Having it before the turn number increment seems contradictory to what you wanted to accomplish (though I'm still vague on what that is exactly)...

Reply:
I suppose it doesn't matter much if techs are actually added to the empire before or after the turn number increment as far as the player is concerned; the number can be tweaked for record keeping purposes either way.

But it is worth considering what will happen if an effect grants an empire a tech. Presumably that should be recorded as being researched on the same turn as a researched tech regardless of the mechanism by which the tech is added. But, I expect that tech-granting effects will fire earlier in the turn sequence, right after the comment `// execute all effects and update meters prior to production, research, etc.`, and will directly add the tech to the empire, before the `CheckResearchProgress` calls even happen (or the subsequent `ApplyMeterEffectsAndUpdateMeters` call and meter back propagation)...

Reply:
So I guess the tech granting effect should in the same way "delay" tech application as the techs granted by research?

Reply:
There'd need to be a list of techs to grant, and the effect would add to that, rather than granting them directly... Similar to effects that directly or indirectly destroy objects, or setting object visibility, which add those objects to lists that are processed later.

Reply:
> What's the reason for adding techs to empires here / now, though?
The important part is that it is later than the  `ApplyMeterEffectsAndUpdateMeters` call because that one will have a different result than the first effect invocation (it will 
A good example is #2555 the first effect invocation happens before the better mine tech is added. So it does the lower tech version and generate the appriopiate sitrep. If the tech is added before the second effect application, it will apply the higher tech version damage but not generate any sitrep. So the player will get the wrong sitrep. 

> Adding the techs is also still before an empire capital check, various visibility and object knowledge tests and empire elimination... Why not later? Having it before the turn number increment seems contradictory to what you wanted to accomplish (though I'm still vague on what that is exactly)...
I guess it could/should be later. I will push the addition of techs after turn increment. 

Tech adding effects are also still on the TODO list for this PR. So is this work-in-progress?

Reply:
Done/Ready-for-review. 
Turn tech researched for initial tech is turn -32768 (on my machine)
For all other tech it is the turn in which the SitRep is shown.

Note GiveTechToEmpire works delayed; this gives an opportunity to add a SitRep to the turn the "Tech researched" message is shown; e.g. "Conquering the planet of the ancient FooBar race provides you with the wisdom of the 'Banana Peeling' technology."